src/dataflow/src/render/mod.rs

```diff
@@ -683,37 +686,134 @@ where
+            (key, row)
+        });
+
+        // Each partition needs to be handled by its own worker, so that we can write messages in order.
+        // For now, we only support single-partition sinks.
+        let keyed = keyed
+            .inner
+            .exchange(move |_| sink_id.hashed())
+            .as_collection();
+
+        // Apply the envelope.
+        // * "Debezium" consolidates the stream, sorts it by time, and produces DiffPairs from it.
+        //   It then renders those as Avro.
+        // * Upsert" does the same, except at the last step, it renders the diff pair in upsert format.
+        //   (As part of doing so, it asserts that there are not multiple conflicting values at the same timestamp)
+        // * "Tail" writes some metadata.
+        let collection = match sink.envelope {
+            SinkEnvelope::Debezium => {
+                let combined = combine_at_timestamp(keyed.arrange_by_key().stream);
+                let rp = Rc::new(RefCell::new(RowPacker::new()));
+                let collection = combined.flat_map(move |(mut k, v)| {
+                    let max_idx = v.len() - 1;
+                    let rp = rp.clone();
+                    v.into_iter().enumerate().map(move |(idx, dp)| {
+                        let k = if idx == max_idx { k.take() } else { k.clone() };
+                        (k, Some(dbz_format(&mut *rp.borrow_mut(), dp)))
+                    })
+                });
+                collection
+            }
+            SinkEnvelope::Upsert => {
+                let combined = combine_at_timestamp(keyed.arrange_by_key().stream);
+
+                let collection = combined.map(|(k, v)| {
+                    let v = upsert_format(v);
+                    (k, v)
+                });
+                collection
+            }
+            SinkEnvelope::Tail { emit_progress } => keyed
+                .inner
+                .map({
+                    let mut rp = RowPacker::new();
+                    move |((k, v), time, diff)| {
+                        rp.push(Datum::Decimal(Significand::new(i128::from(time))));
+                        if emit_progress {
+                            rp.push(Datum::False);
+                        }
+                        rp.push(Datum::Int64(i64::cast_from(diff)));
+                        rp.extend_by_row(&v);
+                        // Add the unpacked timestamp so we can sort by them later.
+                        let v = rp.finish_and_reuse();
+                        ((k, Some(v)), time, 1)
+                    }
+                })
+                .as_collection(),
+        };
+
+        // Some sinks require that the timestamp be appended to the end of the value.
+        let append_timestamp = match &sink.connector {
+            SinkConnector::Kafka(c) => c.consistency.is_some(),
+            SinkConnector::Tail(_) => false,
+            SinkConnector::AvroOcf(_) => false,
+        };
+        let collection = if append_timestamp {
```

Comment:
sorry could you point me to where this is used? i skimmed the encoding changes but theres a bunch there and I didnt see it at first glance

Reply:
It ultimately populates the "transaction" column in Kafka sink output.

E.g. https://github.com/MaterializeInc/materialize/blob/dbe8bc4117b992e29996dd5e1821113f79f2b080/test/testdrive/avro-sinks.td#L198-L201

Reply:
I get that, I just didnt see where in the code that actually happens. i can also do a closer read later

Reply:
Well, the output happens automatically. That's the idea of making sinks "dumb" - everything in the row will just be serialized by kafka.rs .

But, we _do_ have to add something to the schema, for the serialization to work and for the field to be named "transaction"; that happens in `Encoder::new()`: 

```
        if include_transaction {
            // TODO(rkhaitan): this schema omits the total_order and data collection_order
            // fields found in Debezium's transaction metadata struct. We chose to omit
            // those because the order is not stable across reruns and has no semantic
            // meaning for records within a timestamp in Materialize. These fields may
            // be useful in the future for deduplication.
            value_columns.push((
                "transaction".into(),
                ColumnType {
                    nullable: false,
                    scalar_type: ScalarType::Record {
                        fields: vec![(
                            "id".into(),
                            ColumnType {
                                scalar_type: ScalarType::String,
                                nullable: false,
                            },
                        )],
                        custom_oid: Some(
                            custom_types_info.new_named("transaction".to_string()),
                        ),
                    },
                },
            ));
        }
        let writer_schema = build_schema(&value_columns, &custom_types_info);
```